beacon_node/beacon_chain/src/snapshot_cache.rs

```diff
@@ -178,11 +185,29 @@ impl<T: EthSpec> SnapshotCache<T> {
     /// If available, returns a `CacheItem` that should be used for importing/processing a block.
     /// The method will remove the block from `self`, carrying across any caches that may or may not
     /// be built.
-    pub fn get_state_for_block_processing(&mut self, block_root: Hash256) -> Option<CacheItem<T>> {
+    /// In the event the block being processed was observed late, clone the block instead of
+    /// removing it. This allows us to process the next block quickly in the case of a re-org.
+    pub fn get_state_for_block_processing(
+        &mut self,
+        block_root: Hash256,
+        block_delay: Option<Duration>,
+        spec: &ChainSpec,
+    ) -> Option<CacheItem<T>> {
         self.snapshots
             .iter()
             .position(|snapshot| snapshot.beacon_block_root == block_root)
-            .map(|i| self.snapshots.remove(i))
+            .map(|i| {
+                if let Some(delay) = block_delay {
+                    if delay >= MINIMUM_BLOCK_DELAY_FOR_CLONE
+                        && delay <= Duration::from_secs(spec.seconds_per_slot) * 4
+                    {
+                        if let Some(cache) = self.snapshots.get(i) {
+                            return cache.clone();
```

Comment:
I think we may as well avoid cloning the whole `CacheItem` here, as it potentially contains two beacon states. I made a commit on my branch that does this: https://github.com/sigp/lighthouse/pull/2833/commits/f4df466f4f696e37549ca4b3ecf61437f520e3d8

Reply:
Amazing, thanks!